---
title: "Supplementary Figure 1 J"
author: "RAC"
date: "08/05/2021"
output:
  pdf_document: default
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(warning=FALSE, message=FALSE, tidy.opts = list(width.cutoff = 60), tidy = TRUE)

library(dplyr)
library(tibble)
library(tidyr)
library(ggplot2)
library(Hmisc)
library(ggpointdensity)
library(viridis)
```

```{r}
exonIntronReads <- read.csv("../../data/xiCLIP_intronExon.200820.counts", sep="\t", header = F) 
totalCounts <- read.csv("../../data/xiCLIP.read2.totalcounts.200402.tab", sep="\t", header = F)
libraryScalings <- read.csv("../../data/rRNAFactor.tab", sep=" ", header = F)

exonAnno <-read.table("../../data/hg38_HeLa_trimmed_loci_major_primary_isoform_annotated_exon_numbered.bed", header = F)
intronAnno <-read.table("../../data/hg38_HeLa_trimmed_loci_major_primary_isoform_annotated_intron_numbered.bed", header = F)

colnames(exonIntronReads)<-c("Sample","chr","start","end","ID","segmentNumber","strand","count")
colnames(totalCounts)<-c("Sample","TotalCount")
colnames(libraryScalings)<-c("Sample","scaling")

colnames(exonAnno)<-c("chr","start","end","ID","segmentNumber","strand")
colnames(intronAnno)<-c("chr","start","end","ID","segmentNumber","strand")

head(exonIntronReads)
head(totalCounts)
head(libraryScalings)
```

make dataframes with meta data from annotation file
```{r MetaData}

totalExons<-exonAnno %>% 
  separate(ID, into = c("GeneID","Biotype"), sep=":::") %>%
  separate(segmentNumber, into = c("Segment", "SegmentNumber"), sep = "_") %>% 
  group_by(GeneID, Biotype) %>% 
  summarise(TotalExons = max(as.numeric(SegmentNumber))) %>%
  mutate(Exonic = case_when(
    TotalExons > 1 ~ "multiExonic",
    TRUE ~ "monoExonic"
  )) 


tcDf<-totalCounts %>%
  separate(Sample, into=c("Protein","Rep","Timepoint","readType"))

#wrangle count file 
eIRDf<-exonIntronReads %>%
  separate(Sample, into=c("Protein","Rep","Timepoint","readType")) %>%
  separate(ID, into = c("GeneID","Biotype"), sep=":::") %>%
  separate(segmentNumber, into = c("Segment", "SegmentNumber")) %>% 
  filter(!(Protein == "CBP20" & Rep == "3"))



libScale <- libraryScalings %>% 
    separate(Sample, into=c("Protein","Rep","Timepoint")) 

totalCountsDF<-merge(tcDf, libScale)
head(totalCountsDF)
head(eIRDf)
  
```

Supplementary figure 1 j 

plotting log10 CLIP density vs total exons of TU, stratified by protein 
```{r, cache= TRUE}
library(dplyr)
library(tibble)
library(tidyr)
library(ggplot2)
library(Hmisc)
library(ggpointdensity)
library(viridis)

#create dataframe with gene size
gene_size<-
rbind(exonAnno,intronAnno) %>% 
  separate(ID, c("GeneID","biotype"), sep = ":::") %>%
  group_by(GeneID) %>% 
  summarise(gene_size = sum(end-start))


exonvstranscriptdensity<-
eIRDf %>% 
  left_join(totalExons) %>%
  left_join(gene_size) %>% 
  filter(Timepoint == "DMSO"
        & TotalExons > 1 ) %>% #select multiexonic TUs
  group_by(Protein, Rep, GeneID, Timepoint,Segment, TotalExons, gene_size) %>%
  #calculate density of CLIP reads per gene
  summarise(count = sum(count)) %>% 
  mutate(Densitykb = count/(as.numeric(gene_size)/1000)) %>% 
  ungroup() %>%
  left_join(totalCountsDF) %>%
  #normalise to rRNA factor
  mutate(normTorRNAAdj_Densitykb = as.numeric(Densitykb)*as.numeric(scaling),
         countNorm = as.numeric(count) * as.numeric(scaling)) %>% 
  mutate(Timepoint_r = case_when(
    Timepoint == "PBSDRB" ~ "t00",
    TRUE ~ Timepoint
  )) %>% 
  mutate(Timepoint_r = factor(Timepoint_r, levels = c("negative", "t00", "t05", "t10", "t15", "t20", "t40", "t60", "DMSO")))

p<-
exonvstranscriptdensity %>%
filter(Segment == "exon") %>% #select only exonic regions
ggplot(aes(x=as.numeric(TotalExons), y=log2(normTorRNAAdj_Densitykb))) +
  geom_pointdensity(adjust = 5, alpha = 0.25, size =0.1) +
  scale_color_viridis(option = "A",
                      breaks=c(1,40000),
                      labels=c("low","High")) +
  geom_smooth(method="lm") + 
  facet_grid(Timepoint_r~Protein) + 
  coord_equal(ratio = 5)

  p +
  theme_bw() +
  theme(text = element_text(size = 8),
        axis.text.x = element_text(angle = 90, hjust = 1), 
        legend.position = "none") +
  xlab("Total Exons") + 
  ylab("Density (Log2; read/kb)") 


```



